File: algorithms/11-container-of-water/main.py
from typing import List

def max_area_brute_force(height: List[int]) -> int:
    size  = len(height)
    result = 0

    for left in range(size):
        for right in reversed(range(size)):
            print(left, right, height[left], height[right])
            if left >= right:
                continue
            area = (right - left) * min(height[left], height[right])
            if area > result:
                result = area

    return result


def max_area_dp(height: List[int]) -> int:
    size  = len(height)
    result = 0

    left = 0
    right = size - 1

    while left < right:
        print(left, right, height[left], height[right])
        area = (right - left) * min(height[left], height[right])
        if area > result:
            result = area

        if height[left] < height[right]:
            left += 1
        else:
            right -= 1

    return result


def test_max_area_brute_force():
    assert max_area_brute_force([1,8,6,2,5,4,8,3,7]) == 49


def test_max_area_dp():
    assert max_area_dp([1,8,6,2,5,4,8,3,7]) == 49

File: algorithms/11-container-of-water/main.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class Main {
    public static int maxArea(List<Integer> height){
        int size = height.size();

        int left = 0;
        int right = size -1;
        int area = 0;

        while(left < right){
            int newArea = (right - left) * Math.min(height.get(left), height.get(right));
            if(newArea > area){
                area = newArea;
            }
            if(height.get(left) < height.get(right)){
                left += 1;
            } else {
                right -= 1;
            }
        }

        return area;
    }

    public static void main(String[] args){
        List<Integer> list = new ArrayList<>(Arrays.asList(1,8,6,2,5,4,8,3,7));
        int result = maxArea(list);

        assert result == 49 : "Expected 49 but got " + result;
        System.out.println("✅ Test passed!");
    }
}